google/cloud/internal/async_read_write_stream_impl_test.cc

```diff
@@ -0,0 +1,185 @@
+  MOCK_METHOD(void, Read, (FakeResponse*, void*), (override));
+  MOCK_METHOD(void, Write, (FakeRequest const&, void*), (override));
+  MOCK_METHOD(void, Write, (FakeRequest const&, grpc::WriteOptions, void*),
+              (override));
+  MOCK_METHOD(void, Finish, (grpc::Status*, void*), (override));
+  MOCK_METHOD(void, StartCall, (void*), (override));
+  MOCK_METHOD(void, ReadInitialMetadata, (void*), (override));
+};
+
+class MockStub {
+ public:
+  MOCK_METHOD(MockReturnType, FakeRpc,
+              (grpc::ClientContext*, grpc::CompletionQueue*), ());
+};
+
+class MockCompletionQueueImpl : public CompletionQueueImpl {
+ public:
+  MOCK_METHOD0(Run, void());
+  MOCK_METHOD0(Shutdown, void());
+  MOCK_METHOD0(CancelAll, void());
+  MOCK_METHOD1(MakeDeadlineTimer,
+               future<StatusOr<std::chrono::system_clock::time_point>>(
+                   std::chrono::system_clock::time_point));
+  MOCK_METHOD1(MakeRelativeTimer,
+               future<StatusOr<std::chrono::system_clock::time_point>>(
+                   std::chrono::nanoseconds));
+  MOCK_METHOD1(RunAsync, void(std::unique_ptr<RunAsyncBase>));
+
+  MOCK_METHOD2(StartOperation, void(std::shared_ptr<AsyncGrpcOperation>,
+                                    absl::FunctionRef<void(void*)>));
+
+  MOCK_METHOD0(cq, grpc::CompletionQueue&());
+};
+
+TEST(AsyncReadWriteStreamingRpcTest, Basic) {
+  MockStub mock;
+  EXPECT_CALL(mock, FakeRpc)
+      .WillOnce([](grpc::ClientContext*, grpc::CompletionQueue*) {
+        auto stream = absl::make_unique<MockReaderWriter>();
+        EXPECT_CALL(*stream, StartCall).Times(1);
+        EXPECT_CALL(*stream, Write(FakeRequest{"key0"}, _, _)).Times(1);
+        EXPECT_CALL(*stream, Read)
+            .WillOnce([](FakeResponse* response, void*) {
+              response->key = "key0";
+              response->value = "value0_0";
+            })
+            .WillOnce([](FakeResponse* response, void*) {
+              response->key = "key0";
+              response->value = "value0_1";
+            })
+            .WillOnce([](FakeResponse*, void*) {});
+        EXPECT_CALL(*stream, WritesDone).Times(1);
+        EXPECT_CALL(*stream, Finish).WillOnce([](grpc::Status* status, void*) {
+          *status = grpc::Status::OK;
+        });
+        return stream;
+      });
+
+  auto mock_cq = std::make_shared<MockCompletionQueueImpl>();
+  grpc::CompletionQueue grpc_cq;
+  EXPECT_CALL(*mock_cq, cq).WillRepeatedly(ReturnRef(grpc_cq));
+
+  std::deque<std::shared_ptr<AsyncGrpcOperation>> operations;
+  auto notify_next_op = [&](bool ok = true) {
+    auto op = std::move(operations.front());
+    operations.pop_front();
+    op->Notify(ok);
+  };
+
+  EXPECT_CALL(*mock_cq, StartOperation)
+      .WillRepeatedly([&operations](std::shared_ptr<AsyncGrpcOperation> op,
+                                    absl::FunctionRef<void(void*)> call) {
+        operations.push_back(std::move(op));
+        call(op.get());
```

Comment:
Use after move?

Reply:
Ooops, thanks.  It was harmless in this case (not used by name of the mocks called by this mock), but a bad habit nonetheless.